File: WePeiYang/lostfound/src/main/java/com/twtstudio/tjwhm/lostfound/search/SearchActivity.java
package com.twtstudio.tjwhm.lostfound.search;

import android.content.Context;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.LinearLayout;
import android.widget.ProgressBar;

import com.twtstudio.tjwhm.lostfound.R;
import com.twtstudio.tjwhm.lostfound.R2;
import com.twtstudio.tjwhm.lostfound.base.BaseActivity;
import com.twtstudio.tjwhm.lostfound.waterfall.WaterfallBean;
import com.twtstudio.tjwhm.lostfound.waterfall.WaterfallTableAdapter;

import java.util.ArrayList;

import butterknife.BindView;

import static com.facebook.common.internal.Ints.max;

/**
 * Created by tjwhm on 2017/7/2.
 **/

public class SearchActivity extends BaseActivity implements SearchContract.SearchUIView {

    private InputMethodManager inputMethodManager;
    private WaterfallBean waterfallBean;
    private WaterfallTableAdapter tableAdapter;
    private StaggeredGridLayoutManager layoutManager;
    SearchContract.SearchPresenter searchPresenter;

    @BindView(R2.id.search_toolbar)
    Toolbar toolbar;
    @BindView(R2.id.search_searview)
    SearchView searchView;
    @BindView(R2.id.search_recyclerView)
    RecyclerView search_recyclerView;
    @BindView(R2.id.search_progress)
    ProgressBar search_progress;
    @BindView(R2.id.search_no_res)
    LinearLayout search_no_res;

    int page = 1;
    String keyword;
    boolean isLoading = false;


    @Override
    protected int getLayoutResourceId() {
        return R.layout.lf_activity_search;
    }

    @Override
    protected Toolbar getToolbarView() {
        return toolbar;
    }

    @Override
    protected boolean isShowBackArrow() {
        return true;
    }

    @Override
    protected int getToolbarMenu() {
        return 0;
    }

    @Override
    protected void setToolbarMenuClickEvent() {
        toolbar.setNavigationOnClickListener(view -> {
            hideInputKeyboard();
            finish();
        });
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        search_progress.setVisibility(View.GONE);
        search_no_res.setVisibility(View.GONE);
        SearchView.SearchAutoComplete tv = (SearchView.SearchAutoComplete) searchView.findViewById(R.id.search_src_text);
        tv.setTextColor(Color.WHITE);
        tv.setHintTextColor(Color.WHITE);
        searchView.onActionViewExpanded();
        initValues();
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                search_progress.setVisibility(View.VISIBLE);
                waterfallBean.data.clear();
                keyword = query;
                searchPresenter.loadSearchData(keyword, page);
                hideInputKeyboard();
                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                return false;
            }
        });

        search_recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                int totalCount = layoutManager.getItemCount();
                int[] lastPositions = new int[layoutManager.getSpanCount()];
                layoutManager.findLastCompletelyVisibleItemPositions(lastPositions);

                int lastPosition = max(lastPositions);
                if (!isLoading && totalCount < lastPosition + 2 && lastPosition != -1) {
                    ++page;
                    isLoading = true;
                    searchPresenter.loadSearchData(keyword, page);

                }
            }
        });
    }

    private void initValues() {
        waterfallBean = new WaterfallBean();
        waterfallBean.data = new ArrayList<>();
        tableAdapter = new WaterfallTableAdapter(waterfallBean, this,"unknown");
        layoutManager = new StaggeredGridLayoutManager(2, LinearLayoutManager.VERTICAL);
        search_recyclerView.setLayoutManager(layoutManager);
        search_recyclerView.setAdapter(tableAdapter);
        searchPresenter = new SearchPresenterImpl(this);
    }

    private void hideInputKeyboard() {
        inputMethodManager = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        if (inputMethodManager != null) {
            View v = SearchActivity.this.getCurrentFocus();
            if (v == null) {
                return;
            }
            inputMethodManager.hideSoftInputFromWindow(v.getWindowToken(),
                    InputMethodManager.HIDE_NOT_ALWAYS);
            searchView.clearFocus();
        }
    }

    @Override
    public void setSearchData(WaterfallBean waterfallBean) {
        this.waterfallBean.error_code = waterfallBean.error_code;
        this.waterfallBean.message = waterfallBean.message;
        this.waterfallBean.data.addAll(waterfallBean.data);
        tableAdapter.notifyDataSetChanged();
        if (waterfallBean.data.size() == 0 && page == 1) {
            search_no_res.setVisibility(View.VISIBLE);
        } else {
            search_no_res.setVisibility(View.GONE);
        }
        search_progress.setVisibility(View.GONE);
        isLoading = false;
    }

    @Override
    protected void onResume() {
        super.onResume();

    }
}

File: WePeiYang/lostfound2/src/main/java/com/example/lostfond2/search/SearchActivity.kt
package com.example.lostfond2.search

import android.content.Context
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v4.view.MenuItemCompat
import android.support.v7.widget.*
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.Window
import android.view.inputmethod.InputMethodManager
import android.widget.LinearLayout
import android.widget.ProgressBar
import com.example.lostfond2.R
import com.example.lostfond2.service.MyListDataOrSearchBean
import com.example.lostfond2.waterfall.WaterfallTableAdapter


class SearchActivity : AppCompatActivity(), SearchContract.SearchUIView {
    private lateinit var inputMethodManager: InputMethodManager
    private lateinit var waterfallTableAdapter: WaterfallTableAdapter
    private lateinit var waterfallBean: MutableList<MyListDataOrSearchBean>
    private lateinit var layoutManager: StaggeredGridLayoutManager
    private lateinit var searchPresenter: SearchContract.SearchPresenter
    private lateinit var searchView: SearchView
    private lateinit var toolbar: Toolbar
    private lateinit var search_progress: ProgressBar
    private lateinit var recyclerView: RecyclerView
    private lateinit var linearLayour: LinearLayout

    lateinit var keyword: String
    var page = 1
    var isLoading = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        supportRequestWindowFeature(Window.FEATURE_NO_TITLE)//隐藏actionbar，需在setContentView前面
        setContentView(R.layout.activity_search)
        toolbar = findViewById(R.id.search_toolbar)
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        search_progress = findViewById(R.id.search_progress)
        search_progress.visibility = View.GONE
        linearLayour = findViewById(R.id.search_no_res)//绑定的是activity_search的linearlayout
        linearLayour.visibility = View.GONE
        recyclerView = findViewById(R.id.search_searview)
        initValues()
        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                val totalCount = layoutManager.itemCount
                val lastPositions = IntArray(layoutManager.spanCount)
                layoutManager.findLastCompletelyVisibleItemPositions(lastPositions)

                val lastPosition = findMax(lastPositions)
                if (!isLoading && totalCount < lastPosition + 2 && lastPosition != -1) {
                    ++page
                    isLoading = true
                    searchPresenter.loadSearchData(keyword, page)

                }
            }
        })

    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        getMenuInflater().inflate(R.menu.lf_waterfall_menu, menu)
        var searchItem: MenuItem = menu!!.findItem(R.id.waterfall_search)
        //通过MenuItem得到SearchView
        searchView = MenuItemCompat.getActionView(searchItem) as SearchView
        searchView.setIconifiedByDefault(true)//搜索小图标位置在框外
        searchView.setSubmitButtonEnabled(true)
        //获得输入框
        val searchAutoComplete = searchView.findViewById(R.id.search_src_text) as SearchView.SearchAutoComplete
        searchView.setQueryHint("输入卡号/地点/物件名称");
        //设置输入框提示文字样式
        searchAutoComplete.setHintTextColor(getResources().getColor(android.R.color.darker_gray))
        searchAutoComplete.setTextColor(getResources().getColor(android.R.color.background_light))
        //设置搜索框有字时显示叉叉，无字时隐藏叉叉
        searchView.onActionViewExpanded()
        searchView.setIconified(true)

        //监听搜索框文字变化
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String?): Boolean {
                search_progress.visibility = View.VISIBLE
                waterfallBean.clear()
                keyword = query.toString()
                searchPresenter.loadSearchData(keyword, page)
                hideInputKeyboard()
                return true
            }

            override fun onQueryTextChange(newText: String?): Boolean {
                //可以做搜索提示功能
                return false
            }

        })


        return super.onCreateOptionsMenu(menu)

    }


    private fun initValues() {
        waterfallBean = mutableListOf()
        waterfallTableAdapter = WaterfallTableAdapter(waterfallBean, this, "unknown")
        layoutManager = StaggeredGridLayoutManager(2, LinearLayoutManager.VERTICAL)
        recyclerView.setLayoutManager(layoutManager)
        recyclerView.setAdapter(waterfallTableAdapter)
        searchPresenter = SearchPresenterImpl(this)//传啥进去？
    }

    private fun hideInputKeyboard() {
        inputMethodManager = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        if (inputMethodManager != null) {
            val v = this@SearchActivity.currentFocus
            if (v == null) return
            inputMethodManager.hideSoftInputFromWindow(v.windowToken,
                    InputMethodManager.HIDE_NOT_ALWAYS)
            searchView.clearFocus()
        }

    }

    public fun findMax(array: IntArray): Int {
        var maxNumber = array[0]
        val it = array.iterator()
        while (it.hasNext()) {
            val currentNumber = it.nextInt()
            if (currentNumber > maxNumber)
                maxNumber = currentNumber
        }
        return maxNumber
    }

    override fun setSearchData(waterfallBean: MutableList<MyListDataOrSearchBean>) {
        this.waterfallBean = waterfallBean
        waterfallTableAdapter.notifyDataSetChanged()
        if (waterfallBean.size === 0 && page == 1) {
            linearLayour.setVisibility(View.VISIBLE)
        } else {
            linearLayour.setVisibility(View.GONE)
        }
        search_progress.visibility = View.GONE
        isLoading = false
    }

    override fun onResume() {
        super.onResume()
    }
}